4.4 Integrate ExceptionTrap with ErrorHandlerMiddleware

markstory opened the pull request:
Support ExceptionTrap instances. This will let people who need callbacks to be preserved to use `ExceptionTrap::instance()` and config array based configurations to start using ExceptionTrap by default. Defaulting to the new implementation is why there aren't many tests added, as I'm using the existing tests for coverage.

Refs cakephp/cakephp#16228

markstory committed:
4.4 Integrate ExceptionTrap with ErrorHandlerMiddleware

Support ExceptionTrap instances. This will let people who need callbacks
to be preserved to use `ExceptionTrap::instance()` and config array
based configurations to start using ExceptionTrap by default. Defaulting
to the new implementation is why there aren't many tests added, as I'm
using the existing tests for coverage.

Refs cakephp/cakephp#16228

othercorey commented on src/Error/ExceptionTrap.php:
```diff
@@ -235,6 +233,19 @@ public function handleException(Throwable $exception): void
+     * @param \Throwable $exception Exception instance.
+     * @return void
+     */
+    public function applyCallbacks(Throwable $exception): void
```
Is this public just for testing?

markstory replied:
Its public because I needed a way to call this method from the middleware.

othercorey replied:
This doesn't feel great to me. Maybe it's the name. Should we be relying on something less janky like events to run through the error handling steps?

markstory replied:
Yeah, events would be better. I'll open another pull request to remove callbacks and use events instead.

othercorey commented on src/Error/Middleware/ErrorHandlerMiddleware.php:
```diff
@@ -75,10 +76,17 @@ class ErrorHandlerMiddleware implements MiddlewareInterface
+     *
+     * @var \Cake\Error\ExceptionTrap|null
+     */
+    protected $exceptionTrap;
```
Defaulting to null will help with 5.x merge.

othercorey commented on src/Error/Middleware/ErrorHandlerMiddleware.php:
```diff
@@ -201,4 +224,20 @@ protected function getErrorHandler(): ErrorHandler
+        if ($this->exceptionTrap === null) {
+            /** @var class-string<\Cake\Error\ExceptionTrap> $className */
+            $className = App::className('ExceptionTrap', 'Error');
+            $this->exceptionTrap = new $className($this->getConfig());
```
So, won't calling this override the non-trap configured in the constructor?

markstory replied:
It would. I'm relying on this method not being called when there is an `errorHandler` set. I could make that more explicit in this method by raising an error if the errorhandler is defined. What do you think?

othercorey replied:
It's an edge case that most people won't encounter, but since its a public getter, we should probably error.

markstory committed:
Implement interface on integration stub.

markstory committed:
Fix phpcs and linting errors.

markstory committed:
4.4 Remove callbacks from exception/error traps

Use events instead. The event subsystem offers better consistency with
the rest of the framework and provides the necessary hooks for DebugKit
to easily collect errors and exceptions.

markstory committed:
Fix phpcs

markstory committed:
Merge pull request #16290 from cakephp/4next-error-events

4.4 Remove callbacks from exception/error traps
